### ISSUE TITLE
Sendrcv doesn't seem to work across NATs

### ISSUE THREAD

**Issue by riccardocagnasso** [centricular/gstwebrtc-demos#71]
I played with sendrcv a bit to make it work with blackmagic hardware and other stuff. It works fine on LAN but when I try to phone to the other office and say "hey, open https://webrtc.nirbheek.in/ and read me the number" it doesn't. The video doesn't appear to either side of the connection.

The rough schema of the thing is this:

![blank diagram 1](https://user-images.githubusercontent.com/130899/49590501-599c4d80-f96c-11e8-866c-6dd9df9f2b1e.png)

This is what you can see on "the other end"

![screenshot from 2018-12-06 14-57-23](https://user-images.githubusercontent.com/130899/49590541-746ec200-f96c-11e8-90c0-ff07fb006fa2.png)

And the console dump

```Registering with server
webrtc.js:104 Received HELLO
webrtc.js:46 Registered with server, waiting for call
webrtc.js:104 Received {"sdp": {"type": "offer", "sdp": "v=0\r\no=- 7899375669583325737 0 IN IP4 0.0.0.0\r\ns=-\r\nt=0 0\r\na=ice-options:trickle\r\na=msid-semantic:WMS sendrecv\r\nm=video 9 UDP/TLS/RTP/SAVPF 97\r\nc=IN IP4 0.0.0.0\r\na=setup:actpass\r\na=ice-ufrag:afCPuwaaig8PJkk8VP4eCGNpcyWRsGKv\r\na=ice-pwd:xLo+INP9d/++73BqEsljvzU0cK3jvWLb\r\na=sendrecv\r\na=rtcp-mux\r\na=rtcp-rsize\r\na=rtpmap:97 VP8/90000\r\na=rtcp-fb:97 nack pli\r\na=framerate:30\r\na=ssrc:2024735259 msid:user1456566688@host-847dbde4 webrtctransceiver0\r\na=ssrc:2024735259 cname:user1456566688@host-847dbde4\r\na=mid:video0\r\na=fingerprint:sha-256 9C:70:E2:2C:88:72:12:B4:71:66:15:97:FA:25:52:D3:A1:AA:89:0D:02:7C:30:4E:28:E8:A1:2A:B3:71:2E:17\r\nm=audio 9 UDP/TLS/RTP/SAVPF 96\r\nc=IN IP4 0.0.0.0\r\na=setup:actpass\r\na=ice-ufrag:FrcJ2Dnf0zSX3NppsPTN3hGR3CMgc9ca\r\na=ice-pwd:Z6xCBNH73lwRYueOEPTbLPHtcK7pKvCE\r\na=sendrecv\r\na=rtcp-mux\r\na=rtcp-rsize\r\na=rtpmap:96 OPUS/48000/2\r\na=rtcp-fb:96 nack pli\r\na=fmtp:96 sprop-maxcapturerate=48000;sprop-stereo=0\r\na=ssrc:377370790 msid:user1456566688@host-847dbde4 webrtctransceiver1\r\na=ssrc:377370790 cname:user1456566688@host-847dbde4\r\na=mid:audio1\r\na=fingerprint:sha-256 9C:70:E2:2C:88:72:12:B4:71:66:15:97:FA:25:52:D3:A1:AA:89:0D:02:7C:30:4E:28:E8:A1:2A:B3:71:2E:17\r\n"}}
webrtc.js:237 Creating RTCPeerConnection
webrtc.js:169 {"video":true,"audio":true}
webrtc.js:46 Created peer connection for call, waiting for SDP
webrtc.js:222 Incoming stream: 1 video tracks and 1 audio tracks
webrtc.js:46 Remote SDP set
webrtc.js:46 Got SDP offer
webrtc.js:104 Received {"ice": {"candidate": "candidate:1 1 UDP 2013266431 fe80::355c:eb9a:7a2a:db81 47430 typ host", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:2 1 TCP 1015022847 fe80::355c:eb9a:7a2a:db81 9 typ host tcptype active", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:3 1 TCP 1010828543 fe80::355c:eb9a:7a2a:db81 36749 typ host tcptype passive", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:4 1 UDP 2013266430 192.168.1.161 45449 typ host", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:5 1 TCP 1015022079 192.168.1.161 9 typ host tcptype active", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:6 1 TCP 1010827775 192.168.1.161 34529 typ host tcptype passive", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:1 2 UDP 2013266430 fe80::355c:eb9a:7a2a:db81 59591 typ host", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:2 2 TCP 1015022846 fe80::355c:eb9a:7a2a:db81 9 typ host tcptype active", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:3 2 TCP 1010828542 fe80::355c:eb9a:7a2a:db81 46549 typ host tcptype passive", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:4 2 UDP 2013266429 192.168.1.161 36834 typ host", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:5 2 TCP 1015022078 192.168.1.161 9 typ host tcptype active", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:6 2 TCP 1010827774 192.168.1.161 58631 typ host tcptype passive", "sdpMLineIndex": 0}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:1 1 UDP 2013266431 fe80::355c:eb9a:7a2a:db81 51147 typ host", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:2 1 TCP 1015022847 fe80::355c:eb9a:7a2a:db81 9 typ host tcptype active", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:3 1 TCP 1010828543 fe80::355c:eb9a:7a2a:db81 50067 typ host tcptype passive", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:4 1 UDP 2013266430 192.168.1.161 37885 typ host", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:5 1 TCP 1015022079 192.168.1.161 9 typ host tcptype active", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:6 1 TCP 1010827775 192.168.1.161 45001 typ host tcptype passive", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:1 2 UDP 2013266430 fe80::355c:eb9a:7a2a:db81 58372 typ host", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:2 2 TCP 1015022846 fe80::355c:eb9a:7a2a:db81 9 typ host tcptype active", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:3 2 TCP 1010828542 fe80::355c:eb9a:7a2a:db81 41271 typ host tcptype passive", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:4 2 UDP 2013266429 192.168.1.161 46959 typ host", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:5 2 TCP 1015022078 192.168.1.161 9 typ host tcptype active", "sdpMLineIndex": 1}}
webrtc.js:104 Received {"ice": {"candidate": "candidate:6 2 TCP 1010827774 192.168.1.161 60385 typ host tcptype passive", "sdpMLineIndex": 1}}
webrtc.js:243 Adding local stream
webrtc.js:46 Got local stream, creating answer
webrtc.js:89 Got local description: {"type":"answer","sdp":"v=0\r\no=- 3842237536074455630 2 IN IP4 127.0.0.1\r\ns=-\r\nt=0 0\r\na=msid-semantic: WMS ZYaKNRi4iweNo05JZYqCk4EIbX6gXdJJkD7v\r\nm=video 9 UDP/TLS/RTP/SAVPF 97\r\nc=IN IP4 0.0.0.0\r\na=rtcp:9 IN IP4 0.0.0.0\r\na=ice-ufrag:AJWl\r\na=ice-pwd:tvER3Zvuih9MX1aahrOPHTlx\r\na=ice-options:trickle\r\na=fingerprint:sha-256 B4:22:04:AE:AB:37:31:BB:B0:8E:32:46:47:1F:3E:1F:BA:C2:BF:0D:6D:58:05:96:8E:FD:E3:C3:75:B2:5B:76\r\na=setup:active\r\na=mid:video0\r\na=sendrecv\r\na=rtcp-mux\r\na=rtcp-rsize\r\na=rtpmap:97 VP8/90000\r\na=rtcp-fb:97 nack pli\r\na=ssrc:2446360206 cname:NK2v6ppvl1krmZds\r\na=ssrc:2446360206 msid:ZYaKNRi4iweNo05JZYqCk4EIbX6gXdJJkD7v 4fff5126-b0ec-4489-9323-f9492f5a7999\r\na=ssrc:2446360206 mslabel:ZYaKNRi4iweNo05JZYqCk4EIbX6gXdJJkD7v\r\na=ssrc:2446360206 label:4fff5126-b0ec-4489-9323-f9492f5a7999\r\nm=audio 9 UDP/TLS/RTP/SAVPF 96\r\nc=IN IP4 0.0.0.0\r\na=rtcp:9 IN IP4 0.0.0.0\r\na=ice-ufrag:4xts\r\na=ice-pwd:nLPGIUF6meB+fxM/8T0Pyqhb\r\na=ice-options:trickle\r\na=fingerprint:sha-256 B4:22:04:AE:AB:37:31:BB:B0:8E:32:46:47:1F:3E:1F:BA:C2:BF:0D:6D:58:05:96:8E:FD:E3:C3:75:B2:5B:76\r\na=setup:active\r\na=mid:audio1\r\na=sendrecv\r\na=rtcp-mux\r\na=rtpmap:96 OPUS/48000/2\r\na=fmtp:96 minptime=10;useinbandfec=1\r\na=ssrc:1045664249 cname:NK2v6ppvl1krmZds\r\na=ssrc:1045664249 msid:ZYaKNRi4iweNo05JZYqCk4EIbX6gXdJJkD7v 523274f6-87a2-4053-b768-da3003ac63c3\r\na=ssrc:1045664249 mslabel:ZYaKNRi4iweNo05JZYqCk4EIbX6gXdJJkD7v\r\na=ssrc:1045664249 label:523274f6-87a2-4053-b768-da3003ac63c3\r\n"}
webrtc.js:46 Sending SDP answer```

**Comment by alexlapa** [comment]
@riccardocagnasso,

Only host candidates in your log. `srflx` candidates should be gathered by `webrtcbin` and passed to client. But, even if you fix it, there is no guarantees it will work. I recommend adding TURN(Coturn is ok) server or(and) putting your server and Coturn in no NAT env.

UPD: hmm, but, accoring to your log, `onaddstream` is fired, so, i believe, ice connection succeeded. And whats going on in `chrome://webrtc-internals` (`about:webrtc` for FF)?

**Comment by riccardocagnasso** [comment]
@alexlapa 
![screenshot from 2018-12-11 16-26-12](https://user-images.githubusercontent.com/130899/49810533-ba0efe80-fd61-11e8-95e2-d20859f8f06c.png)
[webrtc_internals_dump.zip](https://github.com/centricular/gstwebrtc-demos/files/2668169/webrtc_internals_dump.zip)

**Comment by alexlapa** [comment]
@riccardocagnasso ,

Yep. ice connection failed. 

So. server couldn't establish connection to client, this is usually caused by clients network environment and is a normal thing. Client couldn't establish connection to server because server didn't passed his srflx candidates.

Recommendations from my first comment should do the job.

**Comment by riccardocagnasso** [comment]
@alexlapa Should something like this in webrtc.js be enough? Or some modifications to the gstreamer part are required to use a turn server?

Moreover, is it normal that ice connection fails between two NATted peers without proxy/firewalls and other ugly things? This is a bit meh, because it's quite a common configuration. Would it be possible to integrate UPnP port forwarding in this scenario?

> var rtc_configuration = {iceServers: [{urls: "stun:stun.services.mozilla.com"},
>                                       {urls: "stun:stun.l.google.com:19302"},
> 				      {
> 					  urls: "turn:webrtc.foo.bar:3478",
> 					  username: "foo",
> 					  credential: "bar"
> 				      }]};

**Comment by matthew1000** [comment]
Here's a couple of things I learned about webrtcbin and STUN, which may help:

- The Google STUN server works better than the Mozilla one
- Both client and server needs to have a STUN server configured. If you're using the Python version of this demo, it doesn't set a STUN server, so you'll need to add it.

It that fails, I also managed to get a TURN server working ... I did it using ReTurn, and by declaring it on the server (not in the JS). The format has to be:

```
set_property('turn-server', 'turn://my_name:my_password@my_turn_server_hostname')
```
